### PR TITLE
UCT/API: Introduce exported mkey buffer

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Introduce shared mkey buffer in UCT.

## Why ?

Shared mkey buffer will be packed by `uct_md_mkey_pack_v2`.

## How ?

1. Introduce `UCT_MD_FLAG_SHARED_MKEY` MD capability flag.
2. Update `uct_md_mkey_pack_v2` to support packing shared mkey.
3. Introduce `shared_mkey_packed_size` and `global_id` MD attributes.
4. Update all transports to set `shared_mkey_packed_size` and `global_id`.

### CODE REVIEW

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -695,23 +695,26 @@ typedef enum {
...
-    UCT_MD_FLAG_FIXED      = UCS_BIT(5),  /**< MD supports memory allocation with
-                                               fixed address */
-    UCT_MD_FLAG_RKEY_PTR   = UCS_BIT(6),  /**< MD supports direct access to
-                                               remote memory via a pointer that
-                                               is returned by @ref uct_rkey_ptr */
-    UCT_MD_FLAG_SOCKADDR   = UCS_BIT(7),  /**< MD support for client-server
-                                               connection establishment via
-                                               sockaddr */
-    UCT_MD_FLAG_INVALIDATE = UCS_BIT(8)   /**< MD supports memory invalidation */
+    UCT_MD_FLAG_ALLOC       = UCS_BIT(0),  /**< MD supports memory allocation */
```
Comment (yosefe) [inline src/uct/api/uct.h:698]: move comments before field

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -1435,6 +1438,9 @@ struct uct_md_attr {
     char                     component_name[UCT_COMPONENT_NAME_MAX]; /**< Component name */
     size_t                   rkey_packed_size; /**< Size of buffer needed for packed rkey */
     ucs_cpu_set_t            local_cpus;    /**< Mask of CPUs near the resource */
+    size_t                   shared_mkey_packed_size; /**< Size of buffer needed for packed shared mkey. Valid only if
+                                                           @ref UCT_MD_FLAG_SHARED_MKEY is supported by Memory Domain */
+    uint64_t                 global_id; /**< Global id, UINT64_MAX means that identifier is unknown and should not be used */
```
Comment (yosefe) [inline src/uct/api/uct.h:1443]: move comments before field

File: src/uct/api/v2/uct_v2.h [hunk 3]
```diff
@@ -489,21 +495,22 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
...
- *                          @ref uct_md_query.
+ *                          @ref uct_md_attr_t::rkey_packed_size or
+ *                          @ref uct_md_attr_t::shared_mkey_packed_size, as
+ *                          returned by @ref uct_md_query.
  * @return                  Error code.
  */
 ucs_status_t uct_md_mkey_pack_v2(uct_md_h md, uct_mem_h memh,
                                  const uct_md_mkey_pack_params_t *params,
-                                 void *rkey_buffer);
+                                 void *buffer);
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:513]: rename this in UCT and ops as well

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 4]
```diff
@@ -45,20 +45,22 @@ static ucs_config_field_t uct_cuda_copy_md_config_table[] = {
...
-                                    UCS_BIT(UCS_MEMORY_TYPE_CUDA_MANAGED);
-    md_attr->cap.access_mem_types = UCS_BIT(UCS_MEMORY_TYPE_CUDA) |
-                                    UCS_BIT(UCS_MEMORY_TYPE_CUDA_MANAGED);
-    md_attr->cap.detect_mem_types = UCS_BIT(UCS_MEMORY_TYPE_CUDA) |
-                                    UCS_BIT(UCS_MEMORY_TYPE_CUDA_MANAGED);
-    md_attr->cap.max_alloc        = SIZE_MAX;
-    md_attr->cap.max_reg          = ULONG_MAX;
-    md_attr->rkey_packed_size     = 0;
-    md_attr->reg_cost             = UCS_LINEAR_FUNC_ZERO;
+    md_attr->cap.flags               = UCT_MD_FLAG_REG | UCT_MD_FLAG_ALLOC;
```
Comment (yosefe) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c:48]: introduce uct_md_attr_init()

File: src/uct/api/uct.h (hunk 2 above)
Comment (yosefe) [reply src/uct/api/uct.h:1443]: global_id -> global_id[UCT_MD_GLOBAL_ID_MAX] (256B)

Comment (dmitrygx) [reply src/uct/api/uct.h]: > global_id -> global_id[UCT_MD_GLOBAL_ID_MAX] (256B)

done

File: src/uct/api/v2/uct_v2.h (hunk 3 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:513]: done

File: src/uct/cuda/cuda_copy/cuda_copy_md.c (hunk 4 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_copy/cuda_copy_md.c:48]: @yosefe it seems that if we do it, it will require setting some fileds to initial value and then re-initialize it. is it ok?

Comment (yosefe) [reply src/uct/cuda/cuda_copy/cuda_copy_md.c:48]: yes

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -1419,22 +1435,48 @@ struct uct_listener_params {
...
+    ucs_cpu_set_t     local_cpus;
+
+    /** Size of buffer needed for packed shared mkey. Valid only if
+     *  @ref UCT_MD_FLAG_SHARED_MKEY is supported by Memory Domain */
+    size_t            shared_mkey_packed_size;
 
-    char                     component_name[UCT_COMPONENT_NAME_MAX]; /**< Component name */
-    size_t                   rkey_packed_size; /**< Size of buffer needed for packed rkey */
-    ucs_cpu_set_t            local_cpus;    /**< Mask of CPUs near the resource */
+    /** Global id, '\0' means that an ID is unknown and should not be used */
```
Comment (shamisp) [inline src/uct/api/uct.h:1478]: It seems like all the above changes are unrelated.

Comment (shamisp) [reply src/uct/api/uct.h:1478]: Please describe here what is the global ID - unique device identifier in a system. Can be MAC address or GUID.

File: src/uct/api/v2/uct_v2.h [hunk 6]
```diff
@@ -271,7 +271,13 @@ typedef enum {
      * in order for @ref UCT_MD_MEM_DEREG_FLAG_INVALIDATE flag to function
      * the @ref UCT_MD_MKEY_PACK_FLAG_INVALIDATE flag must be set.
      */
-    UCT_MD_MKEY_PACK_FLAG_INVALIDATE = UCS_BIT(0)
+    UCT_MD_MKEY_PACK_FLAG_INVALIDATE = UCS_BIT(0),
+
+    /**
+     * The flag is used to indicate that the memory region could be accessed
+     * by another process using the same MD to perform UCT operations.
```
Comment (shamisp) [inline src/uct/api/v2/uct_v2.h:278]: another process or another OS image ?

File: src/uct/api/v2/uct_v2.h [hunk 7]
```diff
@@ -489,21 +495,22 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
 /**
  * @ingroup UCT_MD
  *
- * @brief Pack a remote key.
+ * @brief Pack a memory key as a remote or shared one.
```
Comment (shamisp) [inline src/uct/api/v2/uct_v2.h:498]: please add full func description

File: src/uct/api/uct.h (hunk 5 above)
Comment (dmitrygx) [reply src/uct/api/uct.h]: done

File: src/uct/api/v2/uct_v2.h (hunk 6 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h]: by an another process which could run on the same CPU (the same OS image) or am another CPU from SmartNIC

File: src/uct/api/v2/uct_v2.h (hunk 7 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:539]: done

File: src/uct/cuda/cuda_copy/cuda_copy_md.c (hunk 4 above)
Comment (dmitrygx) [reply src/uct/cuda/cuda_copy/cuda_copy_md.c]: done

**Comment by dmitrygx** [comment]
@yosefe @shamisp @brminich @tonycurtis could you review pls?
btw, we should decided on the name of the mkey, we have the following options:
- Shared
- Crossed
- Foreign
- Alias
- Virtual

**Comment by yosefe** [comment]
> @yosefe @shamisp @brminich @tonycurtis could you review pls? btw, we should decided on the name of the mkey, we have the following options:
> 
> * Shared
> * Crossed
> * Foreign
> * Alias
> * Virtual

Maybe "external"?

File: src/uct/api/uct.h [hunk 8]
```diff
@@ -695,23 +695,39 @@ typedef enum {
...
+     *  returned by @ref uct_rkey_ptr */
+    UCT_MD_FLAG_RKEY_PTR    = UCS_BIT(6),
+
+    /** MD supports client-server connection establishment via sockaddr */
+    UCT_MD_FLAG_SOCKADDR    = UCS_BIT(7),
+
+    /** MD supports memory invalidation */
+    UCT_MD_FLAG_INVALIDATE  = UCS_BIT(8),
+
+    /** MD supports shared memory keys with an another process using the same
```
Comment (brminich) [inline src/uct/api/uct.h:728]: no need for `an`

File: src/uct/api/uct.h [hunk 9]
```diff
@@ -1432,23 +1448,52 @@ struct uct_listener_params {
...
+
+        /** Bitmap of memory types that Memory Domain can detect if address
+         *  belongs to it */
+        uint64_t detect_mem_types;
+
+        /** Bitmap of memory types that Memory Domain can allocate memory on */
+        uint64_t alloc_mem_types;
+
+        /** Memory types that Memory Domain can access */
+        uint64_t access_mem_types;
```
Comment (brminich) [inline src/uct/api/uct.h:1474]: unrelated

File: src/uct/api/uct.h [hunk 10]
```diff
@@ -1432,23 +1448,52 @@ struct uct_listener_params {
...
-        uint64_t             detect_mem_types; /**< Bitmap of memory types that Memory Domain can detect if address belongs to it */
-        uint64_t             alloc_mem_types;  /**< Bitmap of memory types that Memory Domain can allocate memory on */
-        uint64_t             access_mem_types; /**< Memory types that Memory Domain can access */
+        /** Maximal allocation size */
+        uint64_t max_alloc;
+
+        /** Maximal registration size */
+        size_t   max_reg;
+
+        /** UCT_MD_FLAG_xx */
```
Comment (brminich) [inline src/uct/api/uct.h:1457]: ??

File: src/uct/api/v2/uct_v2.h [hunk 11]
```diff
@@ -281,7 +281,13 @@ typedef enum {
      * in order for @ref UCT_MD_MEM_DEREG_FLAG_INVALIDATE flag to function
      * the @ref UCT_MD_MKEY_PACK_FLAG_INVALIDATE flag must be set.
      */
-    UCT_MD_MKEY_PACK_FLAG_INVALIDATE = UCS_BIT(0)
+    UCT_MD_MKEY_PACK_FLAG_INVALIDATE = UCS_BIT(0),
+
+    /**
+     * The flag is used to indicate that the memory region could be accessed
+     * by an another process using the same MD to perform UCT operations.
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:288]: no need for `an`

File: src/uct/api/v2/uct_v2.h [hunk 12]
```diff
@@ -530,21 +536,26 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
 /**
  * @ingroup UCT_MD
  *
- * @brief Pack a remote key.
+ * @brief Pack a memory key as a remote or shared one.
+ *
+ * This routine packs a local memory handle registered by @ref uct_md_mem_reg
+ * into a memory buffer which then could be deserialized by a peer and used
+ * in UCP operations.
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:543]: plz remove  referring to UCP

File: src/uct/api/v2/uct_v2.h [hunk 13]
```diff
@@ -530,21 +536,26 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
...
+ * into a memory buffer which then could be deserialized by a peer and used
+ * in UCP operations. 
  *
  * @param [in]  md          Handle to memory domain.
  * @param [in]  memh        Pack a remote key for this memory handle.
  * @param [in]  params      Operation parameters, see @ref
  *                          uct_md_mkey_pack_params_t.
- * @param [out] rkey_buffer Pointer to a buffer to hold the packed remote key.
+ * @param [out] buffer      Pointer to a buffer to hold the packed remote key.
  *                          The size of this buffer has should be at least
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:550]: plz remove `has` (it's an old code, but you already modifying it)

File: src/uct/sm/mm/base/mm_md.c [hunk 14]
```diff
@@ -67,25 +67,26 @@ ucs_status_t uct_mm_seg_new(void *address, size_t length, uct_mm_seg_t **seg_p)
...
-        md_attr->cap.max_alloc    = max_alloc;
+        md_attr->cap.flags    |= UCT_MD_FLAG_ALLOC | UCT_MD_FLAG_FIXED;
+        md_attr->cap.max_alloc = max_alloc;
     }
 
-    memset(&md_attr->local_cpus, 0xff, sizeof(md_attr->local_cpus));
+    if (max_reg > 0) {
+        md_attr->cap.flags  |= UCT_MD_FLAG_REG;
+        md_attr->cap.max_reg = max_reg;
+    }
```
Comment (brminich) [inline src/uct/sm/mm/base/mm_md.c:89]: why needed here? Can just leave it in xpmem

**Comment by dmitrygx** [comment]
> > @yosefe @shamisp @brminich @tonycurtis could you review pls? btw, we should decided on the name of the mkey, we have the following options:
> > 
> > * Shared
> > * Crossed
> > * Foreign
> > * Alias
> > * Virtual
> 
> Maybe "external"?

it seems that all options could be split into two groups:
- suitable for exporter (HOST) and importer (DPU):
shared, virtual, alias
- suitable for either exporter (HOST) or importer (DPU):
crossed - suitable only for DPU (crossing - on HOST), foreign - suitable only for DPU, external - suitable only for DPU

imo, "alias" could mean:
- on HOST: a mkey could be aliased
- on DPU: a mkey is alias of some mkey created on HOST.

**Review by brminich — COMMENTED** [review]
I like the word `shared`

**Comment by yosefe** [comment]
maybe "inter" or "ipc" ?

**Comment by brminich** [comment]
> maybe "inter" or "ipc" ?

we already have `rkey` (stands for **r**emote), which is similar to inter.
ipc is better in this sense

File: src/uct/api/uct.h (hunk 9 above)
Comment (dmitrygx) [reply src/uct/api/uct.h:1474]: we decided to move the comments before the field
https://github.com/openucx/ucx/pull/8422#discussion_r939989576

File: src/uct/api/v2/uct_v2.h (hunk 11 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:288]: done

File: src/uct/api/v2/uct_v2.h (hunk 12 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:543]: done - it should be UCT

File: src/uct/api/v2/uct_v2.h (hunk 13 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:550]: done

File: src/uct/sm/mm/base/mm_md.c (hunk 14 above)
Comment (dmitrygx) [reply src/uct/sm/mm/base/mm_md.c:89]: it is moved to the common code

Comment (brminich) [reply src/uct/sm/mm/base/mm_md.c:89]: why needed, if it is relevant for xpmem only?

**Comment by dmitrygx** [comment]
> > maybe "inter" or "ipc" ?
> 
> we already have `rkey` (stands for **r**emote), which is similar to inter. ipc is better in this sense

hmm, "IPC mkey" sounds great because even though they are on different CPUs, this key could be used by a process to access memory registered by process-exporter

**Comment by dmitrygx** [comment]
> > > maybe "inter" or "ipc" ?
> > 
> > 
> > we already have `rkey` (stands for **r**emote), which is similar to inter. ipc is better in this sense
> 
> "IPC mkey" sounds great because even though processes are on different CPUs, this key could be used by a process to access memory registered by a process-exporter

@shamisp @tonycurtis what do you think about "IPC" name?
Just imagine that "IPC" is used instead of "SHARED" in all places of this PR.

**Comment by tonycurtis** [comment]
> On Aug 16, 2022, at 5:53 PM, dmitrygx ***@***.***> wrote:
> 
> 
> maybe "inter" or "ipc" ?
> 
> we already have rkey (stands for remote), which is similar to inter. ipc is better in this sense
> 
> "IPC mkey" sounds great because even though processes are on different CPUs, this key could be used by a process to access memory registered by a process-exporter
> 
> @shamisp <https://github.com/shamisp> @tonycurtis <https://github.com/tonycurtis> what do you think about "IPC" name?
> Just imagine that "IPC" is used instead of "SHARED" in all places of this PR.
> 

Sounds ok to me,
Tony

File: src/uct/sm/mm/base/mm_md.c (hunk 14 above)
Comment (dmitrygx) [reply src/uct/sm/mm/base/mm_md.c]: I see, removed

**Comment by shamisp** [comment]
Based on discussion with @dmitrygx , he should follow up with Jason to make sure that our naming aligns with potential rdma-core naming. The concern that was raised about IPC naming is that actual behavior of the key does not align with IPC.

**Comment by dmitrygx** [comment]
> Based on discussion with @dmitrygx , he should follow up with Jason to make sure that our naming aligns with potential rdma-core naming. The concern that was raised about IPC naming is that actual behavior of the key does not align with IPC.

@shamisp @yosefe @brminich @tonycurtis 
we discussed Cross-GVMI feature with @jgunthorpe, rdma-core won't introduce it in Verbs.
so, we are free to select any name for our API and make it usable by other SmartNICs, if they will have the same functionality.

File: src/uct/api/uct.h [hunk 15]
```diff
@@ -695,23 +695,39 @@ typedef enum {
...
-    UCT_MD_FLAG_FIXED      = UCS_BIT(5),  /**< MD supports memory allocation with
-                                               fixed address */
-    UCT_MD_FLAG_RKEY_PTR   = UCS_BIT(6),  /**< MD supports direct access to
-                                               remote memory via a pointer that
-                                               is returned by @ref uct_rkey_ptr */
-    UCT_MD_FLAG_SOCKADDR   = UCS_BIT(7),  /**< MD support for client-server
-                                               connection establishment via
-                                               sockaddr */
-    UCT_MD_FLAG_INVALIDATE = UCS_BIT(8)   /**< MD supports memory invalidation */
+        /** MD supports memory allocation */
```
Comment (yosefe) [inline src/uct/api/uct.h:698]: indent

File: src/uct/api/uct.h [hunk 16]
```diff
@@ -695,23 +695,39 @@ typedef enum {
...
+    UCT_MD_FLAG_ALLOC       = UCS_BIT(0),
+
+    /** MD supports memory registration */
+    UCT_MD_FLAG_REG         = UCS_BIT(1),
+
+    /** The transport needs a valid local memory handle for zero-copy
```
Comment (yosefe) [inline src/uct/api/uct.h:704]: should be like this
```
/** 
 * The transport needs a valid local memory handle for zero-copy
 * operations
 */
```

File: src/uct/api/uct.h [hunk 17]
```diff
@@ -695,23 +695,39 @@ typedef enum {
...
+     *  returned by @ref uct_rkey_ptr */
+    UCT_MD_FLAG_RKEY_PTR    = UCS_BIT(6),
+
+    /** MD supports client-server connection establishment via sockaddr */
+    UCT_MD_FLAG_SOCKADDR    = UCS_BIT(7),
+
+    /** MD supports memory invalidation */
+    UCT_MD_FLAG_INVALIDATE  = UCS_BIT(8),
+
+    /** MD supports shared memory keys with another process using the same
```
Comment (yosefe) [inline src/uct/api/uct.h:728]: shared memory keys -> sharing memory keys

File: src/uct/api/uct.h [hunk 18]
```diff
@@ -1432,23 +1448,52 @@ struct uct_listener_params {
...
+        /** Memory Domain capability flags with UCT_MD_FLAG_xx format */
+        uint64_t flags;
+
+        /** Bitmap of memory types that Memory Domain can be registered with */
+        uint64_t reg_mem_types;
+
+        /**< Bitmap of memory types that can be cached for this Memory Domain */
+        uint64_t cache_mem_types;
+
+        /** Bitmap of memory types that Memory Domain can detect if address
```
Comment (yosefe) [inline src/uct/api/uct.h:1466]: format

File: src/uct/api/uct.h [hunk 19]
```diff
@@ -1432,23 +1448,52 @@ struct uct_listener_params {
...
+    /** Size of buffer needed for packed shared mkey. Valid only if
+     *  @ref UCT_MD_FLAG_SHARED_MKEY is supported by Memory Domain */
+    size_t            shared_mkey_packed_size;
 
-    char                     component_name[UCT_COMPONENT_NAME_MAX]; /**< Component name */
-    size_t                   rkey_packed_size; /**< Size of buffer needed for packed rkey */
-    ucs_cpu_set_t            local_cpus;    /**< Mask of CPUs near the resource */
+    /** Global id - unique device identifier in a system (can be MAC address or
+     *  GUID), '\0' means that an ID is unknown and should not be used */
+    char              global_id[UCT_MD_GLOBAL_ID_MAX];
```
Comment (yosefe) [inline src/uct/api/uct.h:1496]: this breaks backward compatibility: application that was compiled with older version will get memory overwrite because its struct is smaller
need to depend on https://github.com/openucx/ucx/pull/8428

File: src/uct/api/v2/uct_v2.h (hunk 6 above)
Comment (yosefe) [reply src/uct/api/v2/uct_v2.h:288]: "using the same MD" -> "using the same device"

**Comment by shamisp** [comment]
> > Based on discussion with @dmitrygx , he should follow up with Jason to make sure that our naming aligns with potential rdma-core naming. The concern that was raised about IPC naming is that actual behavior of the key does not align with IPC.
> 
> @shamisp @yosefe @brminich @tonycurtis we discussed Cross-GVMI feature with @jgunthorpe, rdma-core won't introduce it in Verbs. so, we are free to select any name for our API and make it usable by other SmartNICs, if they will have the same functionality.

I think my final vote will go to "exported". Shared is fine but it can be confused with shared memory especially in the context of  UCX, which has shared memory transports.

**Comment by dmitrygx** [comment]
> > > Based on discussion with @dmitrygx , he should follow up with Jason to make sure that our naming aligns with potential rdma-core naming. The concern that was raised about IPC naming is that actual behavior of the key does not align with IPC.
> > 
> > 
> > @shamisp @yosefe @brminich @tonycurtis we discussed Cross-GVMI feature with @jgunthorpe, rdma-core won't introduce it in Verbs. so, we are free to select any name for our API and make it usable by other SmartNICs, if they will have the same functionality.
> 
> I think my final vote will go to "exported". Shared is fine but it can be confused with shared memory especially in the context of UCX, which has shared memory transports.

ok, I think "exported" should be fine, because in UCP we have the following flow:
- HOST: we create an exported mkey using `ucp_mem_map(EXPORTED, params, &mkey)`
- HOST: we serialize the exported mkey using `ucp_memh_pack(mkey, &exported_mkey_buffer)`
- DPU: create `ucp_mem_map(EXPORTED, params = { exported_mkey_buffer }, &mkey)` - I think `EXPORTED` should be ok here and we don't need to add `IMPORTED`flag.
@yosefe @brminich @tonycurtis

**Comment by yosefe** [comment]
> ok, I think "exported" should be fine, because in UCP we have the following flow:

👍

File: src/uct/api/v2/uct_v2.h [hunk 20]
```diff
@@ -549,43 +555,49 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
...
+    /** Indicate CPUs near the resource. */
+    UCT_MD_ATTR_FIELD_LOCAL_CPUS                = UCS_BIT(12),
 
     /** Indicate size of buffer needed for packed rkey. */
-    UCT_MD_ATTR_FIELD_RKEY_PACKED_SIZE = UCS_BIT(11),
+    UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE = UCS_BIT(13),
 
     /** Indicate CPUs near the resource. */
-    UCT_MD_ATTR_FIELD_LOCAL_CPUS       = UCS_BIT(12)
+    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14)
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:600]: comment on line 599 is irrelevant..

File: src/uct/api/v2/uct_v2.h [hunk 21]
```diff
@@ -549,43 +555,49 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
...
+
+    /** Indicate size of buffer needed for packed rkey. */
+    UCT_MD_ATTR_FIELD_RKEY_PACKED_SIZE          = UCS_BIT(11),
+
+    /** Indicate CPUs near the resource. */
+    UCT_MD_ATTR_FIELD_LOCAL_CPUS                = UCS_BIT(12),
 
     /** Indicate size of buffer needed for packed rkey. */
-    UCT_MD_ATTR_FIELD_RKEY_PACKED_SIZE = UCS_BIT(11),
+    UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE = UCS_BIT(13),
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:597]: line 596 - change packed to exported

File: src/uct/api/v2/uct_v2.h [hunk 22]
```diff
@@ -673,6 +685,18 @@ typedef struct {
...
+
+    /**
+     * Size of buffer needed for packing an exported mkey. Valid only if
+     * @ref UCT_MD_FLAG_EXPORTED_MKEY is supported by Memory Domain
+     */
+    size_t            exported_mkey_packed_size;
+
+    /**
+     * Global id - unique device identifier in a system (can be MAC address or
+     * GUID), '\0' means that an ID is unknown and should not be used
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:697]: is it a byte array or a C-string?

File: src/uct/api/v2/uct_v2.h [hunk 23]
```diff
@@ -690,21 +714,26 @@ ucs_status_t uct_md_query_v2(uct_md_h md, uct_md_attr_v2_t *md_attr);
...
+ *                          The size of this buffer should be at least
+ *                          @ref uct_md_attr_t::rkey_packed_size or
+ *                          @ref uct_md_attr_t::shared_mkey_packed_size, as
+ *                          returned by @ref uct_md_query.
  * @return                  Error code.
  */
 ucs_status_t uct_md_mkey_pack_v2(uct_md_h md, uct_mem_h memh,
                                  const uct_md_mkey_pack_params_t *params,
-                                 void *rkey_buffer);
+                                 void *buffer);
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:736]: maybe "mkey_buffer" ?

File: src/uct/api/v2/uct_v2.h (hunk 21 above)
Comment (brminich) [reply src/uct/api/v2/uct_v2.h:597]: also the comment is copy-pasted from `UCT_MD_ATTR_FIELD_RKEY_PACKED_SIZE`

File: src/uct/api/v2/uct_v2.h [hunk 24]
```diff
@@ -693,8 +693,8 @@ typedef struct {
     size_t            exported_mkey_packed_size;
 
     /**
-     * Global id - unique device identifier in a system (can be MAC address or
-     * GUID), '\0' means that an ID is unknown and should not be used
+     * Global identifier - unique device identifier in a system (can be MAC
+     * address or GUID)
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:697]: still not clear if this is a NULL terminated string..

File: src/uct/api/v2/uct_v2.h [hunk 25]
```diff
@@ -693,8 +693,9 @@ typedef struct {
     size_t            exported_mkey_packed_size;
 
     /**
-     * Global identifier - unique device identifier in a system (can be MAC
-     * address or GUID)
+     * Byte array which hold global unique device identifier (for example, can
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:696]: Byte array that holds a globally unique device identifier (for example, a MAC address or a GUID). If global ...

Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h]: done

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe squashed, could you pls approve?

**Comment by dmitrygx** [comment]
@tonycurtis @shamisp could you review pls?

File: src/uct/api/v2/uct_v2.h [hunk 26]
```diff
@@ -281,7 +281,13 @@ typedef enum {
      * in order for @ref UCT_MD_MEM_DEREG_FLAG_INVALIDATE flag to function
      * the @ref UCT_MD_MKEY_PACK_FLAG_INVALIDATE flag must be set.
      */
-    UCT_MD_MKEY_PACK_FLAG_INVALIDATE = UCS_BIT(0)
+    UCT_MD_MKEY_PACK_FLAG_INVALIDATE = UCS_BIT(0),
+
+    /**
+     * The flag is used to indicate that the memory region could be accessed
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:287]: could -> can

File: src/uct/api/v2/uct_v2.h [hunk 27]
```diff
@@ -549,43 +555,49 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
...
-    UCT_MD_ATTR_FIELD_FLAGS            = UCS_BIT(2),
+    UCT_MD_ATTR_FIELD_FLAGS                     = UCS_BIT(2),
 
     /** Indicate memory types that the MD can register. */
-    UCT_MD_ATTR_FIELD_REG_MEM_TYPES    = UCS_BIT(3),
+    UCT_MD_ATTR_FIELD_REG_MEM_TYPES             = UCS_BIT(3),
 
     /** Indicate memory types that the MD can register. */
-    UCT_MD_ATTR_FIELD_CACHE_MEM_TYPES  = UCS_BIT(4),
+    UCT_MD_ATTR_FIELD_CACHE_MEM_TYPES           = UCS_BIT(4),
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:570]: duplicate comment of UCT_MD_ATTR_FIELD_REG_MEM_TYPES

File: src/uct/api/v2/uct_v2.h [hunk 28]
```diff
@@ -549,43 +555,49 @@ ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
...
 
     /** Indicate component name. */
-    UCT_MD_ATTR_FIELD_COMPONENT_NAME   = UCS_BIT(10),
+    UCT_MD_ATTR_FIELD_COMPONENT_NAME            = UCS_BIT(10),
 
     /** Indicate size of buffer needed for packed rkey. */
-    UCT_MD_ATTR_FIELD_RKEY_PACKED_SIZE = UCS_BIT(11),
+    UCT_MD_ATTR_FIELD_RKEY_PACKED_SIZE          = UCS_BIT(11),
 
     /** Indicate CPUs near the resource. */
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:593]: does "near" need more explanation?

File: src/uct/api/v2/uct_v2.h [hunk 29]
```diff
@@ -673,6 +685,19 @@ typedef struct {
...
+    /**
+     * Size of buffer needed for packing an exported mkey. Valid only if
+     * @ref UCT_MD_FLAG_EXPORTED_MKEY is supported by Memory Domain
+     */
+    size_t            exported_mkey_packed_size;
+
+    /**
+     * Byte array that holds a globally unique device identifier (for example,
+     * a MAC address or a GUID). If global identifiers are equal, it means that
+     * Memory Domains belong to same device.
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:698]: the same

File: src/uct/api/v2/uct_v2.h [hunk 30]
```diff
@@ -690,21 +715,26 @@ ucs_status_t uct_md_query_v2(uct_md_h md, uct_md_attr_v2_t *md_attr);
 /**
  * @ingroup UCT_MD
  *
- * @brief Pack a remote key.
+ * @brief Pack a memory key as a remote or shared one.
+ *
+ * This routine packs a local memory handle registered by @ref uct_md_mem_reg
+ * into a memory buffer which then could be deserialized by a peer and used
```
Comment (tonycurtis) [inline src/uct/api/v2/uct_v2.h:721]: comma before "which"

**Review by tonycurtis — CHANGES_REQUESTED** [review]

File: src/uct/api/v2/uct_v2.h (hunk 26 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:287]: fixed

File: src/uct/api/v2/uct_v2.h (hunk 27 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:570]: fixed

File: src/uct/api/v2/uct_v2.h (hunk 28 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:593]: replaced "near" by "closest"

File: src/uct/api/v2/uct_v2.h (hunk 29 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:698]: done

File: src/uct/api/v2/uct_v2.h (hunk 30 above)
Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:721]: done

**Comment by dmitrygx** [comment]
@tonycurtis your comments have been applied.
@brminich @shamisp @yosefe @tonycurtis could you review pls? no need to wait for CI completion for the next review round, since only comments were fixed.

**Comment by dmitrygx** [comment]
@tonycurtis your comments have been applied.
@brminich @shamisp @yosefe @tonycurtis could you review pls? no need to wait for CI completion for the next review round, since only comments were fixed.

**Comment by yosefe** [comment]
@dmitrygx i think u didnt push

**Review by tonycurtis — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
> @dmitrygx i think u didn't push

yes, forgot...

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
@shamisp @brminich could you take a look pls?

File: src/uct/api/v2/uct_v2.h [hunk 31]
```diff
@@ -670,9 +682,22 @@ typedef struct {
...
 
     /**
-     * Mask of CPUs near the resource.
+     * Mask of CPUs closest to the resource.
      */
     ucs_cpu_set_t     local_cpus;
+
+    /**
+     * Size of buffer needed for packing an exported mkey. Valid only if
+     * @ref UCT_MD_FLAG_EXPORTED_MKEY is supported by Memory Domain
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:691]: minor: `.` is missing in the end

**Review by yosefe — APPROVED** [review]

File: src/uct/api/v2/uct_v2.h [hunk 32]
```diff
@@ -690,21 +715,26 @@ ucs_status_t uct_md_query_v2(uct_md_h md, uct_md_attr_v2_t *md_attr);
...
  *
  * @param [in]  md          Handle to memory domain.
  * @param [in]  memh        Pack a remote key for this memory handle.
  * @param [in]  params      Operation parameters, see @ref
  *                          uct_md_mkey_pack_params_t.
- * @param [out] rkey_buffer Pointer to a buffer to hold the packed remote key.
- *                          The size of this buffer has should be at least
- *                          @ref uct_md_attr_t::rkey_packed_size, as returned by
- *                          @ref uct_md_query.
+ * @param [out] mkey_buffer Pointer to a buffer to hold the packed remote key.
```
Comment (shamisp) [inline src/uct/api/v2/uct_v2.h:728]: remote key -> memory key

Comment (dmitrygx) [reply src/uct/api/v2/uct_v2.h:728]: fixed in https://github.com/openucx/ucx/pull/8538
